kevm-pyk/src/kevm_pyk/kevm.py

```diff
@@ -217,7 +235,70 @@ def add_invariant(cterm: CTerm) -> CTerm:
         constraints.append(mlEqualsTrue(KEVM.range_address(get_cell(config, 'ORIGIN_CELL'))))
         constraints.append(mlEqualsTrue(ltInt(KEVM.size_bytearray(get_cell(config, 'CALLDATA_CELL')), KEVM.pow128())))
 
-        return CTerm(mlAnd([config] + list(unique(constraints))))
+        return CTerm(mlAnd([config] + constraints))
+
+    @staticmethod
+    def extract_branches(cterm: CTerm) -> Iterable[KInner]:
+        config, *constraints = cterm
+        k_cell = get_cell(config, 'K_CELL')
+        jumpi_pattern = KEVM.jumpi_applied(KVariable('###PCOUNT'), KVariable('###COND'))
+        pc_next_pattern = KApply('#pc[_]_EVM_InternalOp_OpCode', [KEVM.jumpi()])
+        branch_pattern = KSequence([jumpi_pattern, pc_next_pattern, KEVM.sharp_execute(), KVariable('###CONTINUATION')])
+        if subst := branch_pattern.match(k_cell):
+            cond = subst['###COND']
+            if cond_subst := KEVM.bool_2_word(KVariable('###BOOL_2_WORD')).match(cond):
+                cond = cond_subst['###BOOL_2_WORD']
+            else:
+                cond = KApply('_==Int_', [cond, intToken(0)])
+            return [mlEqualsTrue(cond), mlEqualsTrue(KApply('notBool_', [cond]))]
+        return []
+
+    @staticmethod
+    def is_terminal(cterm: CTerm) -> bool:
+        config, *_ = cterm
+        k_cell = get_cell(config, 'K_CELL')
+        # <k> #halt </k>
+        if k_cell == KEVM.halt():
+            return True
+        elif type(k_cell) is KSequence:
+            # <k> #halt ~> CONTINUATION </k>
+            if k_cell.arity == 2 and k_cell[0] == KEVM.halt() and type(k_cell[1]) is KVariable:
+                # <callDepth> 0 </callDepth>
+                if get_cell(config, 'CALLDEPTH_CELL') == intToken(0):
+                    return True
+        return False
+
+    @staticmethod
+    def abstract(cterm: CTerm) -> CTerm:
+        term = cterm.kast
+        gas_cell = get_cell(term, 'GAS_CELL')
+        if type(gas_cell) is not KVariable:
+            if not (
+                type(gas_cell) is KApply and gas_cell.label.name == 'infGas' and type(gas_cell.args[0]) is KVariable
+            ):
+                term = remove_constraints_for(['GAS_CELL'], term)
+                if type(gas_cell) is KApply and gas_cell.label.name == 'infGas':
+                    term = set_cell(term, 'GAS_CELL', KEVM.inf_gas(KVariable('GAS_CELL')))
+                else:
+                    term = set_cell(term, 'GAS_CELL', KVariable('GAS_CELL'))
+        # memoryused_cell = get_cell(term, 'MEMORYUSED_CELL')
+        # if type(memoryused_cell) is not KVariable or count_vars(term)[memoryused_cell.name] != 1:
+        #     term = remove_constraints_for(['MEMORYUSED_CELL'], term)
+        #     term = set_cell(term, 'MEMORYUSED_CELL', KVariable('MEMORYUSED_CELL'))
+        wordstack_cell = get_cell(term, 'WORDSTACK_CELL')
+        KApply('.WordStack_EVM-TYPES_WordStack')
```

Comment:
Probably `autoflake` got rid of an unused variable here.

Reply:
Removed.

Reply:
Fixed.